### ISSUE TITLE
Support custom data source like VirtualizedList does BETTER.

### ISSUE THREAD

**Issue by wxsms** [Flipkart/recyclerlistview#498]
Pls do not hurry to close this one. I know there is already an exist topic: #88 
but this is not the same.

in #88 , @naqvitalha saids:

> It is already possible. DataProvider can be extended into something custom. Check the following file: https://github.com/Flipkart/recyclerlistview/blob/master/src/core/dependencies/DataProvider.ts
> Inherit DataProvider and override getDataForIndex(). Start using immutable collection instead of array. Let me know if you need further help.

Thanks for that advice, helps a lot. But I fount out that:

1. override `getDataForIndex` is not enough. Because it has also lots of array operations in `cloneWithRows` function. I have to duplicate all logics in it and replace those with immutable `get`.
2. `newInstance` has to be ovverride too.

something like:

```js
export default class ImmutableDataProvider extends BaseDataProvider {
  getDataForIndex (index) {
    return this.getAllData().get(index);
  }

  cloneWithRows (newData, firstModifiedIndex) {
    const dp = this.newInstance(this.rowHasChanged, this.getStableId);
    const newSize = newData.size;
    const iterCount = Math.min(this.getSize(), newSize);
    if (typeof firstModifiedIndex !== 'number') {
      let i = 0;
      for (i = 0; i < iterCount; i++) {
        if (this.rowHasChanged(this.getAllData().get(i), newData.get(i))) {
          break;
        }
      }
      dp._firstIndexToProcess = i;
    } else {
      dp._firstIndexToProcess = Math.max(Math.min(firstModifiedIndex, this.getAllData().size), 0);
    }
    if (dp._firstIndexToProcess !== this.getAllData().size) {
      dp._requiresDataChangeHandling = true;
    }
    dp._data = newData;
    dp._size = newSize;
    // console.react('cloneWithRows', dp._data, dp._size);
    return dp;
  }

  newInstance (rowHasChanged, getStableId) {
    return new ImmutableDataProvider(rowHasChanged, getStableId);
  }
}
```

These makes the override dirty. It duplicate lots of logics in `BaseDataProvider` which should not be.

And what's more, there're no document about how to do this. People have to dig in to issues for it. I think that immutable data source is common in react world, this should be a common usage and well documented.

**Comment by wxsms** [comment]
Pls let me know if any ideas. happy to make PR too.

**Comment by naqvitalha** [comment]
Yeah I agree. It is not clean. If we abstract out data then it can work. Right now code assumes it to be an array unless you override all of it. Feel free to raise a PR, I'm not sure right now on when I'll be able to improve this.

**Comment by black7375** [comment]
@wxsms @naqvitalha 

from #88 

This is the content of the code I put in as an example.
https://github.com/black7375/React-RxJS-Todo/blob/master/src/Tools/RecyclerProvider.ts

There are some differences..
The representative one is `cloneWithRows`.

- Use the equal of the Immutable.js collection, if the same, use the current size
- Size comparison, small list alignment,
- Returns the first index that meets the criteria.Use `findIndex` to reduce access times to `.get` and take full advantage of optimization

I have confirmed that it works properly if you patch only the type of Immunable.js.
https://github.com/black7375/React-RxJS-Todo/tree/master/patches

**Comment by wxsms** [comment]
@black7375 you may open PR too since you already have an idea. I'm not very good at TS.

**Comment by black7375** [comment]
@wxsms 
I will try it on the weekend or until next week.

**Comment by black7375** [comment]
@naqvitalha  @wxsms 
Hello? I create PR.
Is it possible to review?

**Comment by wxsms** [comment]
em... I think that it's better to support any kind of list, not only Immutable.List. like VirtualizedList of ReactNative does. I've try, but not able to work it out beacause my TS ability is poor.

**Comment by black7375** [comment]
Please understand that I am not familiar with React Native.

In the issue #88 , only the example of Immutable.js is given.
> Thanks for cool library.
> DataProvider supports only array as data source. I want to use it with Immutable.js data structures

Is there an example of another data structure?
If the APIs of the data structures are the same, you only need to add the type, but if not, additional implementation is required.

Currently we need the patch I showed earlier to use another type of library.
If the library allows generics, you can make it extensible using `GenericDataProvider<T, K = keyof T>` instead of `BaseDataProvider`, `ListBaseDataProvider`.
